lib/teleterm/clusters/cluster_file_transfer.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package clusters
+
+import (
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/defaults"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/sshutils/sftp"
+	api "github.com/gravitational/teleport/lib/teleterm/api/protogen/golang/v1"
+	"github.com/gravitational/trace"
+	"io"
+	"os"
+	"time"
+)
+
+func (c *Cluster) TransferFile(request *api.FileTransferRequest, server api.TerminalService_TransferFileServer) error {
+	proxyClient, err := c.clusterClient.ConnectToProxy(server.Context())
+	if err != nil {
+		return err
+	}
+	defer proxyClient.Close()
+
+	var config *sftp.Config
+	var configErr error
+
+	if request.GetDirection() == api.FileTransferDirection_FILE_TRANSFER_DIRECTION_DOWNLOAD {
+		config, configErr = sftp.CreateDownloadConfig(request.GetSource(), request.GetDestination(), sftp.Options{})
+	} else {
+		config, configErr = sftp.CreateUploadConfig([]string{request.GetSource()}, request.GetDestination(), sftp.Options{})
+	}
+
+	if configErr != nil {
+		return trace.Wrap(configErr)
+	}
+
+	config.ProgressWriter = func(fileInfo os.FileInfo) io.Writer {
+		return newGrpcFileTransferProgress(fileInfo.Size(), server)
+	}
+
+	clusterServers, err := proxyClient.FindNodesByFilters(server.Context(), proto.ListResourcesRequest{
+		Namespace: defaults.Namespace,
+	})
+
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	var foundServer types.Server
+	for _, clusterServer := range clusterServers {
+		if clusterServer.GetName() == request.GetServerId() {
+			foundServer = clusterServer
+			break
+		}
+	}
+
+	if foundServer == nil {
+		return trace.BadParameter("Requested server does not exist")
+	}
+
+	err = c.clusterClient.TransferFiles(server.Context(), request.GetLogin(), foundServer.GetHostname()+":0", config)
```

Comment:
Shouldn't this be wrapped in `addMetadataToRetryableError` and then in `retryWithRelogin` on the frontend? The call to `proxyClient.FindNodesByFilters` should be within the callback too but I understand you want to find a way to get rid of it.

Reply:
Yes, it should! 
For now I wrapped whole function in `addMetadataToRetryableError` to have frontend part working, I'd like to get rid `proxyClient.FindNodesByFilters` if possible.

https://user-images.githubusercontent.com/20583051/194033799-3cca3373-14eb-404a-8089-47d4e7427c15.mov